sefaria/model/linker/ref_resolver.py

```diff
@@ -435,28 +625,56 @@ def _get_unrefined_ref_part_matches_for_title_context(self, context_ref: Optiona
             matches += [match]
         return matches
 
-    def _apply_context_swaps(self, raw_ref: RawRef, context_swap_map: Dict[str, str]=None):
-        """
-        Use `context_swap_map` to swap matching element of `ref_parts`
-        Allows us to redefine how a ref part is interpreted depending on the context
-        E.g. some rishonim refer to other rishonim based on nicknames
+    def _collect_context_mutations(self, book_context_ref: Optional[text.Ref]) -> Optional[ContextMutationSet]:
+        if book_context_ref is None:
+            return None
+        node = getattr(book_context_ref, 'index_node', None)
+        if node is None:
+            node = getattr(book_context_ref.index, 'nodes', None)
+        if node is None:
+            return None
+        path_nodes = []
+        curr_node = node
+        while curr_node is not None:
+            path_nodes.append(curr_node)
+            curr_node = getattr(curr_node, 'parent', None)
+        mutation_set = ContextMutationSet()
+        for path_node in reversed(path_nodes):
+            raw_mutations = getattr(path_node, 'ref_resolver_context_mutations', None)
+            if not raw_mutations:
+                continue
+            parsed_mutations = self._parse_context_mutation_data(raw_mutations, path_node)
+            if parsed_mutations:
+                mutation_set.add_mutations(parsed_mutations)
+        return mutation_set if len(mutation_set) > 0 else None
 
-        Modifies `raw_ref` with updated ref_parts
-        """
-        swapped_ref_parts = []
-        term_matcher = self.get_term_matcher()
-        if context_swap_map is None: return
-        for part in raw_ref.raw_ref_parts:
-            # TODO assumes only one match in term_matches
-            term_matches = term_matcher.match_term(part)
-            found_match = False
-            for match in term_matches:
-                if match.slug not in context_swap_map: continue
-                swapped_ref_parts += [TermContext(schema.NonUniqueTerm.init(slug)) for slug in context_swap_map[match.slug]]
-                found_match = True
-                break
-            if not found_match: swapped_ref_parts += [part]
-        raw_ref.parts_to_match = swapped_ref_parts
+    @staticmethod
+    def _parse_context_mutation_data(raw_mutations: Iterable[dict], node: schema.SchemaNode) -> List[ContextMutation]:
```

Comment:
why is this function on RefResolver? It looks like it belongs on SchemaNode since it's parsing a field on SchemaNode.